### PR TITLE
StocardImporter: use ZipFile (fixes spanned archive marker breaking import)

### LINKED ISSUE #2096: Stocard / Klarna APP import failed
Google Pixel 6
Android 13 (TQ3A.230901.001.C1)
Catima (Google Play) 2.31.1

`protect.card_locker.FormatException: Couldn't find any loyalty cards in this Stocard export.`

Screenshot_20241002-083235

Stocard is no longer a thing. It will force prompt user to download Klarna APP instead.
I contacted support@stocardapp.com to get my export data. I emailed the `.zip` file and password to catima.g9ex3@hackerchick.me Please take a look.

I checked issue #863, #1242, #1277. Maybe Stocard/Klarna (recently) changed the exported `.zip` structure again, so Catima import failed.

### PR DESCRIPTION
**Pull request by obfusk**
Properly fixes #2096.

~~Tested with this copy of `stocard2.zip` modified to have the same spanned archive marker: [stocard2s.zip](https://github.com/user-attachments/files/17469859/stocard2s.zip).~~

~~Sorting by offset is almost certainly not needed. But some ZIP files have a different ordering of entries in the central directory. This ensures the iteration order is the same as before the change, which may or may not be correct.~~

The other importers should probably also use `ZipFile`. But that can be done one at a time.

### CODE REVIEW

**Review by TheLastProject — APPROVED** [review]
Seems to work well, nice!

Perhaps it would be good to replace the unit test file with your new file (how did you create this?) to make sure the tests actually test a more true-to-live export file (a multi-volume archive).

**Comment by obfusk** [comment]
Using `ZipFile` instead of `ZipInputStream.getNextEntry()` is how ZIP files are meant to processed: by reading the central directory at the end of the file and then using that to look up the local entries. ZIP files aren't really supposed to be processed as streams, though it can be useful to process them like that sometimes. Just gets complicated when there's e.g. a spanned archive marker at the start instead of a local file header as ZIP files can technically contain random data anywhere before/after ZIP entries (which is kind of how you can add an APK v2 signature without breaking anything). So assuming no unexpected regressions I would recommend always using this approach, also for the other importers.

NB: I didn't test this on older Android versions, just 13/14 (don't recall if I updated my emulator to 14 yet). I also didn't test importing a "real" export with the marker in Catima (I did test reading one using `ZipFile` in a standalone script and did confirm the only unexpected thing about that ZIP file was the presence of that marker).

> how did you create this?

I used [my work in progress `ziptools.py`](https://github.com/obfusk/reproducible-apk-tools/blob/ziptools/ziptools.py) to copy it with a prepended marker:

```py
from repro_apk import ziptools
with ziptools.ZipFile.open("stocard2.zip") as zf:
    with ziptools.ZipFile.build("stocard2s.zip") as zb:
        zb._fh.write(0x08074b50.to_bytes(4, "little"))
        zb.copy_from(zf)
```

That also works for encrypted ZIP files (`ziptools.py` can't read encrypted data yet, but it can copy it without modifying any (meta)data).

> a multi-volume archive

I don't think it actually is? It's only one file after all. Just has that marker. But yeah, might be useful to test with that.

File: app/src/main/java/protect/card_locker/importexport/StocardImporter.java
Comment (TheLastProject) [reply app/src/main/java/protect/card_locker/importexport/StocardImporter.java:158]: Would probably be cleaner, to be honest. It used to get closed explicitly before this change.

Comment (obfusk) [reply app/src/main/java/protect/card_locker/importexport/StocardImporter.java:158]: That was a different stream: one for reading everything, not one per entry. Closing that after reading one entry would have been wrong. But yeah, likely this one should be closed after reading (assuming they are indeed independent streams opened per entry). Will need to test that, the documentation doesn't really say.

**Comment by obfusk** [comment]
I'll test closing the stream later. Will update the MR after that, also to update the test ZIP file. I think that's it? Apart from doing more testing which I hope you can do (I don't have older android emulators set up)?

File: app/src/main/java/protect/card_locker/importexport/StocardImporter.java
Comment (obfusk) [inline app/src/main/java/protect/card_locker/importexport/StocardImporter.java:154]: I could also remove this. It should not be needed and deviates from how most programs would read the entries (but only if the order actually differs, which is rare). But since the importer may depend on the current iteration order of the entries I figured it best not to change it unless you want me to.

**Comment by obfusk** [comment]
- updated both stocard ZIP files to include the spanned archive marker
- confirmed both stocard tests failed with the old code
- updated the PR to
  - close the stream
  - not sort the entries (having confirmed the order would not have changed for the export I tested with earlier): this should be correct and if it does end up causing regressions it would probably be better to find out how the code unintentionally depends on the order of processing the entries and make adjustments for that instead
- confirmed both stocard tests still work with the old ZIP files with the new code
- tested importing the new `stocard2.zip` on android 14 in the emulator as well

which leaves:

- further testing (older android, real exports) I hope you can do, maybe some of the affected users could test as well
- updating the other importers too (outside the scope of this PR)

**Comment by obfusk** [comment]
Should I make a follow-up PR to update the other importers too (and clean up a few more things I noticed) after this is merged?

**Comment by TheLastProject** [comment]
> Should I make a follow-up PR to update the other importers too (and clean up a few more things I noticed) after this is merged?

Sure

> Edit: I assume you'll make the changes to the pkpass parser yourself?

Feel free to use GitHub suggestions to suggest the change you'd make. I'll manually add it to the commit (sorry about the lack of commit credit) to keep the history clean.

> further testing (older android, real exports) I hope you can do, maybe some of the affected users could test as well

I'll try to test the real exports I have when I can, but it can't really be worse than the current state (which is broken), so I probably won't spend time to try a lot of different Android releases (don't really have the time).

Will merge after I had time to test, thanks.

**Comment by TheLastProject** [comment]
Can confirm all the exports I got import fine now :)

**Comment by TheLastProject** [comment]
Oh yeah, before I forget: should we document this change in the zip4j issue? Documenting this will probably help others and I guess you could argue it's not really a zip4j bug, at most a documentation issue, unless I misunderstood? Well, bug or not, documenting it for others is probably helpful

**Pull request merged**
